

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Read Through Transcription &mdash; GC_bioinfo 1.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Sequence from Around Max TSS" href="sequence_from_region_around_max_tss.html" />
    <link rel="prev" title="Pausing Distance Distribution from Max TSS" href="pausing_distance_distribution_from_maxTSS.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> GC_bioinfo
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="TES_combined_heatmap.html"><em>TES Combined Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="TES_fold_change_heatmap.html"><em>TES Fold Change Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="TES_heatmap.html"><em>TES Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="base_distribution.html"><em>Base Distribution</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="divergent_pileup_metaplot.html"><em>Divergent Pileup Metaplot</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="five_prime_metaplot.html"><em>Five Prime Metaplot</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_combined_heatmap.html"><em>Gene Body Combined Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_fold_change_heatmap.html"><em>Gene Body Fold Change Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_heatmap.html"><em>Gene Body Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="make_regions_file_centered_on_max_tss.html"><em>Make Regions File Centered on Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="multicoverage.html"><em>Multicoverage</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="nucleotide_heatmap.html"><em>Nucleotide Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="pausing_distance_distribution_from_maxTSS.html"><em>Pausing Distance Distribution from Max TSS</em></a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#"><em>Read Through Transcription</em></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#usage-and-option-summary">Usage and option summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="#behavior">Behavior</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sequence_from_region_around_max_tss.html"><em>Sequence from Around Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="sequence_searches.html"><em>Sequence Searches</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="three_prime_metaplot.html"><em>Three Prime Metaplot</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="tps_distance_per_gene.html"><em>TPS Distance Per Gene</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="track_links_from_bw.html"><em>Track Links From Bigwigs</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="truQuant.html"><em>truQuant</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="tsrFinder.html"><em>tsrFinder</em></a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GC_bioinfo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><em>Read Through Transcription</em></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="read-through-transcription">
<h1><em>Read Through Transcription</em><a class="headerlink" href="#read-through-transcription" title="Permalink to this headline">¶</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">read_through_transcription</span></code> tool computes the coverage of 3’ ends of sequencing data around the 3’ end of
features provided. This can be used to look at read through or runaway transcription by providing the gene body regions
file from truQuant.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This tool requires <a class="reference external" href="https://github.com/arq5x/bedtools2">bedtools</a> to be installed.</p>
</div>
<div class="section" id="usage-and-option-summary">
<h2>Usage and option summary<a class="headerlink" href="#usage-and-option-summary" title="Permalink to this headline">¶</a></h2>
<p><strong>Usage</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GC_bioinfo</span> <span class="n">read_through_transcription</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="p">[</span><span class="n">threads</span><span class="p">]]</span>
                                           <span class="n">regions_filename</span> <span class="n">tsr_file</span>
                                           <span class="n">upstream_distance</span>
                                           <span class="n">downstream_distance</span> <span class="n">interval_size</span>
                                           <span class="n">sequencing_files</span>
                                           <span class="p">[</span><span class="n">sequencing_files</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%" />
<col style="width: 85%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Regions Filename</strong></p></td>
<td><p>Bed formatted file containing all the genes to quantify (regions will be determined from the 3’ end of each region in this file.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>TSR Filename</strong></p></td>
<td><p><a class="reference external" href="https://geoffscollins.github.io/GC_bioinfo/tsrFinder.html">tsrFinder</a> output file which will be blacklisted. Simply type <em>no</em> to not blacklist TSRs.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Upstream Distance</strong></p></td>
<td><p>The number of base pairs to subtract from the left position.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Downstream Distance</strong></p></td>
<td><p>The number of base pairs to add from the left position.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Interval Distance</strong></p></td>
<td><p>The size of sub-regions to split the regions into.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Sequencing Files</strong></p></td>
<td><p>Sequencing files to quantify separated by spaces.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="behavior">
<h2>Behavior<a class="headerlink" href="#behavior" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">read_through_transcription</span></code> will report the position relative to the 3’ end of the regions provided and the sum
of the 3’ reads at that interval.</p>
<p>For example:</p>
<img alt="_images/read_through_transcription.png" src="_images/read_through_transcription.png" />
<p></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ head regions_centered_on_max_tss.bed
chr1    <span class="m">959251</span>  <span class="m">959261</span>  NOC2L   <span class="m">46</span>      -
chr1    <span class="m">960627</span>  <span class="m">960637</span>  KLHL17  <span class="m">27</span>      +
chr1    <span class="m">966516</span>  <span class="m">966526</span>  PLEKHN1 <span class="m">8</span>       +
chr1    <span class="m">1000092</span> <span class="m">1000102</span> HES4    <span class="m">87</span>      -
chr1    <span class="m">1000290</span> <span class="m">1000300</span> ISG15   <span class="m">12</span>      +
chr1    <span class="m">1020114</span> <span class="m">1020124</span> AGRN    <span class="m">35</span>      +
chr1    <span class="m">1074302</span> <span class="m">1074312</span> RNF223  <span class="m">10</span>      -
chr1    <span class="m">1116102</span> <span class="m">1116112</span> C1orf159        <span class="m">9</span>       -
chr1    <span class="m">1231967</span> <span class="m">1231977</span> SDF4    <span class="m">321</span>     -
chr1    <span class="m">1232237</span> <span class="m">1232247</span> B3GALT6 <span class="m">174</span>     +

$ head seq_file-blacklisted_150_20_30_600-TSR.tab
chr1    <span class="m">199198</span>  <span class="m">199348</span>  <span class="m">759</span>     <span class="m">22</span>      +       <span class="m">199260</span>  <span class="m">199261</span>  <span class="m">2</span>       <span class="m">199258</span>
chr1    <span class="m">628994</span>  <span class="m">629144</span>  <span class="m">604</span>     <span class="m">20</span>      +       <span class="m">629092</span>  <span class="m">629093</span>  <span class="m">2</span>       <span class="m">629110</span>
chr1    <span class="m">629431</span>  <span class="m">629581</span>  <span class="m">17821</span>   <span class="m">380</span>     +       <span class="m">629571</span>  <span class="m">629572</span>  <span class="m">188</span>     <span class="m">629551</span>
chr1    <span class="m">629651</span>  <span class="m">629801</span>  <span class="m">6747</span>    <span class="m">206</span>     +       <span class="m">629764</span>  <span class="m">629765</span>  <span class="m">25</span>      <span class="m">629740</span>
chr1    <span class="m">629900</span>  <span class="m">630050</span>  <span class="m">2345</span>    <span class="m">69</span>      +       <span class="m">629929</span>  <span class="m">629930</span>  <span class="m">6</span>       <span class="m">629955</span>
chr1    <span class="m">630309</span>  <span class="m">630459</span>  <span class="m">1511</span>    <span class="m">48</span>      +       <span class="m">630437</span>  <span class="m">630438</span>  <span class="m">7</span>       <span class="m">630406</span>
chr1    <span class="m">630604</span>  <span class="m">630754</span>  <span class="m">3435</span>    <span class="m">80</span>      +       <span class="m">630681</span>  <span class="m">630682</span>  <span class="m">21</span>      <span class="m">630690</span>
chr1    <span class="m">630780</span>  <span class="m">630930</span>  <span class="m">4601</span>    <span class="m">113</span>     +       <span class="m">630893</span>  <span class="m">630894</span>  <span class="m">16</span>      <span class="m">630873</span>
chr1    <span class="m">630999</span>  <span class="m">631149</span>  <span class="m">2722</span>    <span class="m">68</span>      +       <span class="m">631136</span>  <span class="m">631137</span>  <span class="m">6</span>       <span class="m">631116</span>
chr1    <span class="m">631204</span>  <span class="m">631354</span>  <span class="m">2843</span>    <span class="m">85</span>      +       <span class="m">631294</span>  <span class="m">631295</span>  <span class="m">11</span>      <span class="m">631287</span>

$ head seq_file.bed
chr1    <span class="m">11981</span>   <span class="m">12023</span>   A00876:119:HW5F5DRXX:1:2168:2248:1407   <span class="m">255</span>     -
chr1    <span class="m">13099</span>   <span class="m">13117</span>   A00876:119:HW5F5DRXX:1:2203:31403:26757 <span class="m">255</span>     -
chr1    <span class="m">13356</span>   <span class="m">13423</span>   A00876:119:HW5F5DRXX:1:2151:15808:7827  <span class="m">255</span>     -
chr1    <span class="m">13435</span>   <span class="m">13477</span>   A00876:119:HW5F5DRXX:1:2273:15781:19241 <span class="m">255</span>     -
chr1    <span class="m">13739</span>   <span class="m">13772</span>   A00876:119:HW5F5DRXX:1:2256:29966:10520 <span class="m">255</span>     -
chr1    <span class="m">13741</span>   <span class="m">13773</span>   A00876:119:HW5F5DRXX:1:2235:4101:11882  <span class="m">255</span>     -
chr1    <span class="m">14178</span>   <span class="m">14203</span>   A00876:119:HW5F5DRXX:1:2115:8241:31422  <span class="m">255</span>     -
chr1    <span class="m">14734</span>   <span class="m">14768</span>   A00876:119:HW5F5DRXX:1:2165:23764:2440  <span class="m">255</span>     -
chr1    <span class="m">14988</span>   <span class="m">15012</span>   A00876:119:HW5F5DRXX:1:2219:16134:32784 <span class="m">255</span>     -
chr1    <span class="m">18337</span>   <span class="m">18362</span>   A00876:119:HW5F5DRXX:1:2149:32054:31328 <span class="m">255</span>     -

$ GC_bioinfo read_through_transcription seq_file-blacklisted_150_20_30_600-gene_body_regions.bed <span class="se">\</span>
  seq_file-blacklisted_150_20_30_600-TSR.tab <span class="m">1000</span> <span class="m">1000</span> <span class="m">100</span> seq_file.bed <span class="p">&amp;</span>
Position       seq_file.bed
-1000   <span class="m">1924</span>
-900    <span class="m">1808</span>
-800    <span class="m">1762</span>
-700    <span class="m">1695</span>
-600    <span class="m">1807</span>
-500    <span class="m">1856</span>
-400    <span class="m">2999</span>
-300    <span class="m">3838</span>
-200    <span class="m">2055</span>
-100    <span class="m">1911</span>
<span class="m">0</span>       <span class="m">1809</span>
<span class="m">100</span>     <span class="m">1878</span>
<span class="m">200</span>     <span class="m">1977</span>
<span class="m">300</span>     <span class="m">1905</span>
<span class="m">400</span>     <span class="m">1885</span>
<span class="m">500</span>     <span class="m">1890</span>
<span class="m">600</span>     <span class="m">1894</span>
<span class="m">700</span>     <span class="m">1790</span>
<span class="m">800</span>     <span class="m">1718</span>
<span class="m">900</span>     <span class="m">1740</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="sequence_from_region_around_max_tss.html" class="btn btn-neutral float-right" title="Sequence from Around Max TSS" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="pausing_distance_distribution_from_maxTSS.html" class="btn btn-neutral float-left" title="Pausing Distance Distribution from Max TSS" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Geoff Collins.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>