

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>TPS Distance Per Gene &mdash; GC_bioinfo 1.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Track Links From Bigwigs" href="track_links_from_bw.html" />
    <link rel="prev" title="Three Prime Metaplot" href="three_prime_metaplot.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> GC_bioinfo
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="TES_combined_heatmap.html"><em>TES Combined Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="TES_fold_change_heatmap.html"><em>TES Fold Change Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="TES_heatmap.html"><em>TES Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="base_distribution.html"><em>Base Distribution</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="divergent_pileup_metaplot.html"><em>Divergent Pileup Metaplot</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="five_prime_metaplot.html"><em>Five Prime Metaplot</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_combined_heatmap.html"><em>Gene Body Combined Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_fold_change_heatmap.html"><em>Gene Body Fold Change Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_heatmap.html"><em>Gene Body Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="make_regions_file_centered_on_max_tss.html"><em>Make Regions File Centered on Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="multicoverage.html"><em>Multicoverage</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="nucleotide_heatmap.html"><em>Nucleotide Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="pausing_distance_distribution_from_maxTSS.html"><em>Pausing Distance Distribution from Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="read_through_transcription.html"><em>Read Through Transcription</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="sequence_from_region_around_max_tss.html"><em>Sequence from Around Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="sequence_searches.html"><em>Sequence Searches</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="three_prime_metaplot.html"><em>Three Prime Metaplot</em></a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#"><em>TPS Distance Per Gene</em></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#behavior">Behavior</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="track_links_from_bw.html"><em>Track Links From Bigwigs</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="truQuant.html"><em>truQuant</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="tsrFinder.html"><em>tsrFinder</em></a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GC_bioinfo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><em>TPS Distance Per Gene</em></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tps-distance-per-gene">
<h1><em>TPS Distance Per Gene</em><a class="headerlink" href="#tps-distance-per-gene" title="Permalink to this headline">¶</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">tps_distance_per_gene</span></code> tool computes the most commonly used transcript length for each region in the region file centered on the max TSS.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This tool requires <a class="reference external" href="https://github.com/arq5x/bedtools2">bedtools</a> to be installed.</p>
</div>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p><strong>Usage</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GC_bioinfo</span> <span class="n">tps_distance_per_gene</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="p">[</span><span class="n">threads</span><span class="p">]]</span>
                                      <span class="n">regions_filename</span> <span class="n">sequencing_files</span>
                                      <span class="p">[</span><span class="n">sequencing_files</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%" />
<col style="width: 85%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Regions filename</strong></p></td>
<td><p>Bed formatted file containing all the regions you want to quantify (must be centered on +1 nt). This file can be generated from the
<a class="reference external" href="https://geoffscollins.github.io/GC_bioinfo/make_regions_file_centered_on_max_tss.html">make_regions_file_centered_on_max_tss program</a></p></td>
</tr>
<tr class="row-odd"><td><p><strong>Sequencing Files</strong></p></td>
<td><p>Bed formatted file from a sequencing experiment.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%" />
<col style="width: 85%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>-t, –threads</strong></p></td>
<td><p>Maximum number of threads. Default is the number of threads on the system. This program will not use more threads than the number of sequencing files provided</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Sequencing Files</strong></p></td>
<td><p>Additional sequencing files can be provided.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="behavior">
<h2>Behavior<a class="headerlink" href="#behavior" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">tps_distance_per_gene</span></code> will report the most commonly used transcript length from the maxTSS (regions file centered) for each gene</p>
<p>For example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ head seq_file.bed
chr1    <span class="m">11981</span>   <span class="m">12023</span>   A00876:119:HW5F5DRXX:1:2168:2248:1407   <span class="m">255</span>     -
chr1    <span class="m">13099</span>   <span class="m">13117</span>   A00876:119:HW5F5DRXX:1:2203:31403:26757 <span class="m">255</span>     -
chr1    <span class="m">13356</span>   <span class="m">13423</span>   A00876:119:HW5F5DRXX:1:2151:15808:7827  <span class="m">255</span>     -
chr1    <span class="m">13435</span>   <span class="m">13477</span>   A00876:119:HW5F5DRXX:1:2273:15781:19241 <span class="m">255</span>     -
chr1    <span class="m">13739</span>   <span class="m">13772</span>   A00876:119:HW5F5DRXX:1:2256:29966:10520 <span class="m">255</span>     -
chr1    <span class="m">13741</span>   <span class="m">13773</span>   A00876:119:HW5F5DRXX:1:2235:4101:11882  <span class="m">255</span>     -
chr1    <span class="m">14178</span>   <span class="m">14203</span>   A00876:119:HW5F5DRXX:1:2115:8241:31422  <span class="m">255</span>     -
chr1    <span class="m">14734</span>   <span class="m">14768</span>   A00876:119:HW5F5DRXX:1:2165:23764:2440  <span class="m">255</span>     -
chr1    <span class="m">14988</span>   <span class="m">15012</span>   A00876:119:HW5F5DRXX:1:2219:16134:32784 <span class="m">255</span>     -
chr1    <span class="m">18337</span>   <span class="m">18362</span>   A00876:119:HW5F5DRXX:1:2149:32054:31328 <span class="m">255</span>     -

$ head regions_centered_on_max_tss.bed
chr1    <span class="m">959251</span>  <span class="m">959261</span>  NOC2L   <span class="m">46</span>      -
chr1    <span class="m">960627</span>  <span class="m">960637</span>  KLHL17  <span class="m">27</span>      +
chr1    <span class="m">966516</span>  <span class="m">966526</span>  PLEKHN1 <span class="m">8</span>       +
chr1    <span class="m">1000092</span> <span class="m">1000102</span> HES4    <span class="m">87</span>      -
chr1    <span class="m">1000290</span> <span class="m">1000300</span> ISG15   <span class="m">12</span>      +
chr1    <span class="m">1020114</span> <span class="m">1020124</span> AGRN    <span class="m">35</span>      +
chr1    <span class="m">1074302</span> <span class="m">1074312</span> RNF223  <span class="m">10</span>      -
chr1    <span class="m">1116102</span> <span class="m">1116112</span> C1orf159        <span class="m">9</span>       -
chr1    <span class="m">1231967</span> <span class="m">1231977</span> SDF4    <span class="m">321</span>     -
chr1    <span class="m">1232237</span> <span class="m">1232247</span> B3GALT6 <span class="m">174</span>     +

$ GC_bioinfo tps_distance_per_gene regions_centered_on_max_tss.bed seq_file.bed &gt; seq_file_tps_distance_per_gene.txt
$ head seq_file_tps_distance_per_gene.txt
Gene    seq_file.bed
NOC2L   <span class="m">32</span>
KLHL17  <span class="m">42</span>
PLEKHN1 <span class="m">45</span>
HES4    <span class="m">16</span>
ISG15   <span class="m">44</span>
AGRN    <span class="m">33</span>
RNF223  <span class="m">35</span>
C1orf159        N/A
SDF4    <span class="m">68</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="track_links_from_bw.html" class="btn btn-neutral float-right" title="Track Links From Bigwigs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="three_prime_metaplot.html" class="btn btn-neutral float-left" title="Three Prime Metaplot" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Geoff Collins.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>