

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>tsrFinder &mdash; GC_bioinfo 1.0.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="truQuant" href="truQuant.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> GC_bioinfo
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="TES_fold_change_heatmap.html"><em>TES Fold Change Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="TES_heatmap.html"><em>TES Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="base_distribution.html"><em>Base Distribution</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_fold_change_heatmap.html"><em>Gene Body Fold Change Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="gene_body_heatmap.html"><em>Gene Body Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="make_regions_file_centered_on_max_tss.html"><em>Make Regions File Centered on Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="metaplot.html"><em>Metaplot</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="multicoverage.html"><em>Multicoverage</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="nucleotide_heatmap.html"><em>Nucleotide Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="pausing_distance_distribution_from_maxTSS.html"><em>Pausing Distance Distribution from Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="read_through_transcription.html"><em>Read Through Transcription</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="region_fold_change_heatmap.html"><em>Region Fold Change Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="region_heatmap.html"><em>Region Heatmap</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="sequence_from_region_around_max_tss.html"><em>Sequence from Around Max TSS</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="sequence_searches.html"><em>Sequence Searches</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="tps_distance_per_gene.html"><em>TPS Distance Per Gene</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="track_links_from_bw.html"><em>Track Links From Bigwigs</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="truQuant.html"><em>truQuant</em></a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#"><em>tsrFinder</em></a><ul>
<li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#behavior">Behavior</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GC_bioinfo</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><em>tsrFinder</em></li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tsrfinder">
<h1><em>tsrFinder</em><a class="headerlink" href="#tsrfinder" title="Permalink to this headline">¶</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">tsrFinder</span></code> tool identifies transcription start regions from PRO-Cap, PRO-Seq, and related sequencing experiments.
This tsrFinder is a much quicker version (with few output modifications) of the previous two tsrFinders: <a class="reference external" href="https://github.com/P-TEFb/tsrFinder">one</a> and
<a class="reference external" href="https://github.com/P-TEFb/tsrFinderM1">two</a>.</p>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p><strong>Usage</strong>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GC_bioinfo</span> <span class="n">tsrFinder</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="p">[</span><span class="n">threads</span><span class="p">]]</span>
                          <span class="n">seq_file</span> <span class="n">window_size</span> <span class="n">min_seq_depth</span>
                          <span class="n">min_avg_transcript_length</span> <span class="n">max_fragment_size</span>
                          <span class="n">chrom_size_file</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 83%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Required Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Sequencing File</strong></p></td>
<td><p>Bed formatted file from a sequencing experiment.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Window Size</strong></p></td>
<td><p>The size of the TSRs to find.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Min Seq Depth</strong></p></td>
<td><p>The minimum number of 5’ reads to be considered as a TSR.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Min avg Transcript Length</strong></p></td>
<td><p>The minimum average transcript length will eliminate TSRs from sequencing artifacts.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Max fragment size</strong></p></td>
<td><p>The maximum transcript length for a read to be included in tsrFinder analysis.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Chrom size file</strong></p></td>
<td><p>A file containing the chromosome sizes. This can be optained using <a class="reference external" href="http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/">fetchChromSizes</a>
The hg38 chromsome size file can be found in the GC_bioinfo static directory.</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 15%" />
<col style="width: 85%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Optional Arguments</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>-t, –threads</strong></p></td>
<td><p>Maximum number of threads. Default is the number of threads on the system. This program will not use more threads than twice the number of chromosomes in the
chrom size file.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="behavior">
<h2>Behavior<a class="headerlink" href="#behavior" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">tsrFinder</span></code> generate a file named the sequencing file plus the tsrFinder parameters and -TSR.bed. This file contains
the chromosome, left position, right position, the read sum, number of 5’ ends, strand, TSS left and right positions,
the TSS strength, and the position of the average TSS (weighted mean location).</p>
<p>For example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$ head seq_file.bed
chr1    <span class="m">11981</span>   <span class="m">12023</span>   A00876:119:HW5F5DRXX:1:2168:2248:1407   <span class="m">255</span>     -
chr1    <span class="m">13099</span>   <span class="m">13117</span>   A00876:119:HW5F5DRXX:1:2203:31403:26757 <span class="m">255</span>     -
chr1    <span class="m">13356</span>   <span class="m">13423</span>   A00876:119:HW5F5DRXX:1:2151:15808:7827  <span class="m">255</span>     -
chr1    <span class="m">13435</span>   <span class="m">13477</span>   A00876:119:HW5F5DRXX:1:2273:15781:19241 <span class="m">255</span>     -
chr1    <span class="m">13739</span>   <span class="m">13772</span>   A00876:119:HW5F5DRXX:1:2256:29966:10520 <span class="m">255</span>     -
chr1    <span class="m">13741</span>   <span class="m">13773</span>   A00876:119:HW5F5DRXX:1:2235:4101:11882  <span class="m">255</span>     -
chr1    <span class="m">14178</span>   <span class="m">14203</span>   A00876:119:HW5F5DRXX:1:2115:8241:31422  <span class="m">255</span>     -
chr1    <span class="m">14734</span>   <span class="m">14768</span>   A00876:119:HW5F5DRXX:1:2165:23764:2440  <span class="m">255</span>     -
chr1    <span class="m">14988</span>   <span class="m">15012</span>   A00876:119:HW5F5DRXX:1:2219:16134:32784 <span class="m">255</span>     -
chr1    <span class="m">18337</span>   <span class="m">18362</span>   A00876:119:HW5F5DRXX:1:2149:32054:31328 <span class="m">255</span>     -

$ head hg38.chrom.sizes
chr1    <span class="m">248956422</span>
chr2    <span class="m">242193529</span>
chr3    <span class="m">198295559</span>
chr4    <span class="m">190214555</span>
chr5    <span class="m">181538259</span>
chr6    <span class="m">170805979</span>
chr7    <span class="m">159345973</span>
chrX    <span class="m">156040895</span>
chr8    <span class="m">145138636</span>
chr9    <span class="m">138394717</span>

$ GC_bioinfo tsrFinder seq_file.bed <span class="m">20</span> <span class="m">20</span> <span class="m">30</span> <span class="m">600</span> hg38.chrom.sizes
$ head seq_file_20_20_30_600-TSR.tab
chr1    <span class="m">629421</span>  <span class="m">629441</span>  <span class="m">738</span>     <span class="m">20</span>      +       <span class="m">629431</span>  <span class="m">629432</span>  <span class="m">4</span>       <span class="m">629431</span>
chr1    <span class="m">629490</span>  <span class="m">629510</span>  <span class="m">2263</span>    <span class="m">64</span>      +       <span class="m">629494</span>  <span class="m">629495</span>  <span class="m">14</span>      <span class="m">629500</span>
chr1    <span class="m">629564</span>  <span class="m">629584</span>  <span class="m">13877</span>   <span class="m">273</span>     +       <span class="m">629571</span>  <span class="m">629572</span>  <span class="m">188</span>     <span class="m">629572</span>
chr1    <span class="m">629685</span>  <span class="m">629705</span>  <span class="m">939</span>     <span class="m">21</span>      +       <span class="m">629698</span>  <span class="m">629699</span>  <span class="m">6</span>       <span class="m">629697</span>
chr1    <span class="m">629708</span>  <span class="m">629728</span>  <span class="m">1701</span>    <span class="m">43</span>      +       <span class="m">629723</span>  <span class="m">629724</span>  <span class="m">12</span>      <span class="m">629719</span>
chr1    <span class="m">629740</span>  <span class="m">629760</span>  <span class="m">1911</span>    <span class="m">63</span>      +       <span class="m">629759</span>  <span class="m">629760</span>  <span class="m">21</span>      <span class="m">629753</span>
chr1    <span class="m">629919</span>  <span class="m">629939</span>  <span class="m">1219</span>    <span class="m">32</span>      +       <span class="m">629929</span>  <span class="m">629930</span>  <span class="m">6</span>       <span class="m">629931</span>
chr1    <span class="m">630666</span>  <span class="m">630686</span>  <span class="m">2277</span>    <span class="m">44</span>      +       <span class="m">630681</span>  <span class="m">630682</span>  <span class="m">21</span>      <span class="m">630679</span>
chr1    <span class="m">630824</span>  <span class="m">630844</span>  <span class="m">658</span>     <span class="m">21</span>      +       <span class="m">630828</span>  <span class="m">630829</span>  <span class="m">4</span>       <span class="m">630834</span>
chr1    <span class="m">630879</span>  <span class="m">630899</span>  <span class="m">2237</span>    <span class="m">49</span>      +       <span class="m">630893</span>  <span class="m">630894</span>  <span class="m">16</span>      <span class="m">630890</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="truQuant.html" class="btn btn-neutral float-left" title="truQuant" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Geoff Collins.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>